src/aggregator/client/queue.go

```diff
@@ -214,15 +219,26 @@ func (q *queue) Flush() {
 		n   int
 		err error
 	)
+	// ensure Flush is safe for concurrent use
+	q.processingMtx.Lock()
+	defer q.processingMtx.Unlock()
+
+	// replace incoming buffer with a fresh one, to have a snapshot of queue before flush and unblock new writes
+	q.mtx.Lock()
+	q.bufProcessing, q.bufIncoming = q.bufIncoming, q.bufProcessing
+	q.mtx.Unlock()
 
 	for err == nil {
 		// flush everything in batches, to make sure no single payload is too large,
 		// to prevent a) allocs and b) timeouts due to big buffer IO taking too long.
 		var processed int
-		processed, err = q.flush(buf)
+		processed, err = q.flush(&q.bufProcessing, buf)
 		n += processed
 	}
 
+	// drop any unconsumed messages
+	q.bufProcessing.reset()
```

Comment:
Sanity checking: 

- Do we have/need retries before we drop the buffer here?
- Are these buffered elsewhere? Afaict no, but sanity checking.